src/com/google/enterprise/adaptor/database/UniqueKey.java

```diff
@@ -160,6 +161,57 @@
     return aclSqlCols;
   }
 
+  /**
+   * Attempt to extract ColumnTypes from the supplied Map
+   * of column names to java.sql.Type
+   *
+   * @param sqlTypes Map of column names to Integer java.sql.Types
+   */
+  void addColumnTypes(Map<String, Integer> sqlTypes) {
```

Comment:
can this be something that is part of constructor?  

Adding this first mutator makes this class no longer immutable.  

Database adaptor could replace its instance of UniqueKey, rather than modify its unique key, maybe?